android-core/src/main/kotlin/com/mparticle/internal/ConfigManager.kt

```diff
@@ -0,0 +1,1436 @@
+        }
+
+    open val userStorage: UserStorage?
+        get() = getUserStorage(mpid)
+
+    fun getUserStorage(mpId: Long): UserStorage? {
+        if (mUserStorage == null || mUserStorage?.mpid != mpId) {
+            mUserStorage = mContext?.let { create(it, mpId) }
+        }
+        return mUserStorage
+    }
+
+    fun deleteUserStorage(context: Context, mpid: Long) {
+        if (mUserStorage != null) {
+            mUserStorage?.deleteUserConfig(context, mpid)
+        }
+    }
+
+    fun deleteUserStorage(mpId: Long) {
+        mContext?.let { deleteUserStorage(it, mpId) }
+    }
+
+    fun migrateConfigIfNeeded() {
+        if (sPreferences?.getBoolean(MIGRATED_TO_KIT_SHARED_PREFS, false) == false) {
+            sPreferences?.edit()?.putBoolean(MIGRATED_TO_KIT_SHARED_PREFS, true)?.apply()
+            val configString: String? = sPreferences?.getString(CONFIG_JSON, null)
+            if (!isEmpty(configString)) {
+                try {
+                    // save ourselves some time and only parse the JSONObject if might contain the embedded kits key
+                    if (configString?.contains("\"" + KEY_EMBEDDED_KITS + "\":") == true) {
+                        Logger.info("Migrating kit configuration")
+                        saveConfigJson(JSONObject(configString), etag, ifModified, configTimestamp)
+                    }
+                } catch (jse: JSONException) {
+                }
+            }
+        }
+    }
+
+    /**
+     * detrmine if the stored config age is greater than mMaxConfigAge and clear it from storage if it is
+     */
+    fun checkConfigStaleness() {
+        var storageDate: Long? = configTimestamp
+        if (storageDate == null) {
+            // migration step: if the current config does not have a timestamp, set one to the current time
+            configTimestamp = System.currentTimeMillis()
+            storageDate = configTimestamp
+        }
+        mMaxConfigAge?.let {
+            if (it < 0) {
+                return
+            }
+        } ?: return
+        if (mMaxConfigAge == 0) {
+            clearConfig()
+        } else {
+            val currentTime = System.currentTimeMillis()
+            val storageDateMillis = storageDate ?: 0L
+            if (currentTime >= storageDateMillis + TimeUnit.SECONDS.toMillis(mMaxConfigAge?.toLong() ?: 0L)) {
+                clearConfig()
+            }
+        }
+    }
+
+    val config: String?
+        get() = sPreferences?.getString(CONFIG_JSON, "")
+
+    var configTimestamp: Long?
+        get() {
+            if (sPreferences?.contains(CONFIG_JSON_TIMESTAMP) == true) {
+                return sPreferences?.getLong(CONFIG_JSON_TIMESTAMP, 0)
+            } else {
+                return null
+            }
+        }
+        set(timestamp) {
+            if (timestamp != null) {
+                sPreferences?.edit()
+                    ?.putLong(CONFIG_JSON_TIMESTAMP, timestamp)
+                    ?.apply()
+            }
```

Comment:
Isn't the logic here different from the original Java? In the original, `putLong` could be called with a null value.

The `putLong` method is defined as: `public abstract SharedPreferences.Editor putLong (String key, long value)` in Java, with `value` being the `long` value type. In the previous logic, if `timestamp` was null, would this have converted that to a `0` and stored it?

In the new Kotlin implementation, it will leave the previously stored value which may not be the desired logic.

Reply:
Note that I haven't tested this myself to confirm, so there may be no difference in the implementations in reality, but worth double checking.